### ISSUE TITLE
dayGridMonth dynamic start / end dates

### ISSUE THREAD

**Issue by Kaihogan** [fullcalendar/fullcalendar#4981]
I have made every effort to try to locate this request on my own from reviewing all of the past issues / request reported, going to Stack Overview, even posting a question there but getting the expected answer that i was fearing.  That the current FullCalendar does not yet offer this feature.    

I did see that this very request was asked 2 years ago and was quickly marked as resolved with a reverence to use the Custom View feature.  

I have explored that as well the only issue is that all of the examples reference everything but the dayGridMonth.   Assuming that those were merely examples I gave it a try to create a custom view with alternate month date ranges.   However, all attempts failed.  

I can therefor only summarize that this one view, the one i need to modify, is not currently a feature of the dayGridMonth view.    Below are a list of the months and their modified date range that i am needing for my purpose.

- January = 01/01 to 01/30
- February = 01/31 to 03/01
-  March = 03/02 to 03/31
-  April = 04/01 to 05/01
-  May = 05/02 to 06/01
-  June = 06/02 to 07/01
-  July = 07/02 to 07/31 
-  August = 08/01 to 08/30
-  September = 08/31 to 09/30
-  October = 10/01 to 10/31
-  November = 11/01 to 12/01
-  December = 12/02 to 12/31

As the dayGridMonth view appears to be the only view without the ability to provide your own date range is this something that can be added?

**Comment by acerix** [comment]
If this is a question, please refer to the [support page](https://fullcalendar.io/support/) and use [Stack Overflow](https://stackoverflow.com/questions/tagged/fullcalendar) for help.  If this is a bug, please supply a [runnable, stripped-down demonstration](http://fullcalendar.io/wiki/Reduced-Test-Cases/).

You can use the `dayGrid` view with a `visibleRange`:

https://codepen.io/acerix/pen/YzKqqwz

**Comment by Kaihogan** [comment]
Thank you for the feedback but sadly what you have proposed will not work for my needs.  In the example that you provided on codepen looks good on the initial start, however, if i do nothing more than change the the viewrange date from 01/01/2019 to 12/31/2019 to one of my examples of 08/31/2019 to 09/30/2019 I no longer see a calendar but rather a non-calendar day grid.   So it still does not appear to do what i am looking for.   It is possible that it should work and this is really no a feature request but a bug report.   Not sure

**Comment by acerix** [comment]
I can see it only works when `visibleRange.start` is Jan 1, the view is completely broken with other dates so that seems like a bug.

https://codepen.io/acerix/pen/mdbEVQg

**Comment by Kaihogan** [comment]
Dylan, thanks for reviewing and confirming regarding this defect.

**Comment by Kaihogan** [comment]
I did a little more testing with the issse and it seems to be specific to full weeks. (starting SUN, ending SAT).   For example if i use the following date range 01/06/2019 to 02/03/2019 then i get again a full calendar because 06JAN is a Sunday and 02FEB is a Saturday.   In addition there is another bug, per se, in this same area.  When entering a date range of 01/06/2019 to 02/03/2019 the expected behavior is a calendar that starts on 06JAN and ends on 03FEB however the calendar will start correctly on 06JAN but then ends one day short (02FEB) of the requested end date.

**Comment by Kaihogan** [comment]
Just to close the loop on this from my needs, now that I understand where the defects are with this feature (at least until a later version is codded to address) I was able to create a customView for my unique month start / stop dates making certain to also have adjustment logic to add days before and at the end of the calendar so that I would Start on a Sunday and end on a Saturday (using a Sunday date).  I know have the calendar working for my needs.

**Comment by gorley** [comment]
Can you please share your customview code.

**Comment by Kaihogan** [comment]
Certainly.   I will show the entire view but more than likely you only need a few lines from this view depending on your needs.  I am currently using fullCalendar version 4.3.1

`			contractMonth: {
      		type: 'dayGrid',
				buttonText: 'month',
				visibleRange: function(currentDate) {
			   	startDate = new Date(currentDate.getFullYear(),currentDate.getMonth(),1);
			   	endDate = new Date(currentDate.getFullYear(),currentDate.getMonth()+1,0);
					monthTitle = monthNames[startDate.getMonth()]+" "+startDate.getFullYear();

					var contractMonthArray = [
						{"Name" : "January","MonthStart" : 1,"DayStart" : 1,"MonthEnd" : 1,DayEnd : 30},
						{"Name" : "February","MonthStart" : 1,"DayStart" : 31,"MonthEnd" : 3,DayEnd : 1},
						{"Name" : "March","MonthStart" : 3,"DayStart" : 2,"MonthEnd" : 3,DayEnd : 31},
						{"Name" : "April","MonthStart" : 4,"DayStart" : 1,"MonthEnd" : 5,DayEnd : 1},
						{"Name" : "May","MonthStart" : 5,"DayStart" : 2,"MonthEnd" : 6,DayEnd : 1},
						{"Name" : "June","MonthStart" : 6,"DayStart" : 2,"MonthEnd" : 7,DayEnd : 1},
						{"Name" : "July","MonthStart" : 7,"DayStart" : 2,"MonthEnd" : 7,DayEnd : 31},
						{"Name" : "August","MonthStart" : 8,"DayStart" : 1,"MonthEnd" : 8,DayEnd : 30},
						{"Name" : "September","MonthStart" : 8,"DayStart" : 31,"MonthEnd" : 9,DayEnd : 30},
						{"Name" : "October","MonthStart" : 10,"DayStart" : 1,"MonthEnd" : 10,DayEnd : 31},
						{"Name" : "November","MonthStart" : 11,"DayStart" : 1,"MonthEnd" : 12,DayEnd : 1},
						{"Name" : "December","MonthStart" : 12,"DayStart" : 2,"MonthEnd" : 12,DayEnd : 31}
					];

					contractMonthArray.forEach(function(e) {
						arrayStartDate = new Date(e.MonthStart+"/"+e.DayStart+"/"+currentDate.getFullYear());
						arrayEndDate = new Date(e.MonthEnd+"/"+e.DayEnd+"/"+currentDate.getFullYear());
						if (currentDate >= arrayStartDate && currentDate <= arrayEndDate)
						{
							startDate = arrayStartDate;
							endDate = arrayEndDate;
							monthTitle = e.Name + " " + currentDate.getFullYear();
						}
					});
					// Adjust the start & end dates, respectively to address a code bug in the FullCalendar
			   	startDateChk = new Date(startDate);
			   	endDateChk = new Date(endDate);
				   startDate.setDate(startDate.getDate() - (startDate.getDay() - 0  )  ); // back to last Sunday
			   	endDate.setDate(endDate.getDate() + (6 - endDate.getDay())+1); // Forward to next Sunday.  There is a bug in current code that will show the end date minus 1 day (Saturday).
			    	return { start: startDate, end: endDate };
				}
			}
		},
`

**Comment by Azaferany** [comment]
@acerix 
any update on this?
its still breaks on 6.1.8
